### PR TITLE
p2p: plumb rudamentary service discovery to rectors and update statesync

### PR DESCRIPTION
**Pull request by tychoish**
This is a little coarse, but the idea is that we'll send information
about the channels a peer has upon the peer-up event that we send to
reactors that we can then use to reject peers (if neeeded) from reactors.

This solves the problem where statesync would hang in test networks
(and presumably real) where we would attempt to statesync from seed
nodes, thereby hanging silently forever.

### CODE REVIEW

File: internal/p2p/p2ptest/require.go [hunk 1]
```diff
@@ -136,6 +136,7 @@ func RequireUpdate(t *testing.T, peerUpdates *p2p.PeerUpdates, expect p2p.PeerUp
 
 	select {
 	case update := <-peerUpdates.Updates():
+		update.Channels = nil
```
Comment (creachadair) [inline internal/p2p/p2ptest/require.go:139]: This I presume is just so we compare correctly independent of the contents of this field?

File: internal/p2p/peermanager.go [hunk 2]
```diff
@@ -678,15 +679,16 @@ func (m *PeerManager) Accepted(peerID types.NodeID) error {
 // peer must already be marked as connected. This is separate from Dialed() and
 // Accepted() to allow the router to set up its internal queues before reactors
 // start sending messages.
```
Comment (creachadair) [inline internal/p2p/peermanager.go:681]: Please add some text about `channels` here too.

File: internal/p2p/peermanager_test.go [hunk 3]
```diff
@@ -1311,7 +1311,7 @@ func TestPeerManager_Ready(t *testing.T) {
 	require.Equal(t, p2p.PeerStatusDown, peerManager.Status(a.NodeID))
 
 	// Marking a as ready should transition it to PeerStatusUp and send an update.
-	peerManager.Ready(ctx, a.NodeID)
+	peerManager.Ready(ctx, a.NodeID, nil)
```
Comment (creachadair) [inline internal/p2p/peermanager_test.go:1314]: Is there a practical way to verify that nontrivial channel maps are also correctly propagated?

**Review by creachadair — COMMENTED** [review]
This seems like a reasonable addition. One question I have is whether we can get into a case where the peer manager thinks one of the channels isn't available incorrectly? I.e., do the channels ever come up at different times, or are they all-or-nothing when the connection is made?

If it's the latter: Perhaps we don't need to track them independently, but could assign a flag: "If all these channels are available, usableForStateSync=true" and then we don't have to check for set membership everywhere?

File: internal/p2p/p2ptest/require.go (hunk 1 above)
Comment (tychoish) [reply internal/p2p/p2ptest/require.go:139]: yeah, I've refactored this so that we just compare the node id and the status.

File: internal/p2p/peermanager_test.go (hunk 3 above)
Comment (tychoish) [reply internal/p2p/peermanager_test.go:1314]: added a unittest

File: .github/workflows/lint.yml [hunk 4]
```diff
@@ -20,7 +20,7 @@ jobs:
             **/**.go
             go.mod
             go.sum
-      - uses: golangci/golangci-lint-action@v3.1.0
+      - uses: golangci/golangci-lint-action@v2.5.2
```
Comment (creachadair) [inline .github/workflows/lint.yml:23]: I'm fine with reverting this, but we should probably do it in a separate PR.

File: internal/p2p/p2ptest/require.go (hunk 1 above)
Comment (creachadair) [inline internal/p2p/p2ptest/require.go:139]: I think perhaps we no longer need to set this nil?

**Review by creachadair — APPROVED** [review]

File: .github/workflows/lint.yml (hunk 4 above)
Comment (tychoish) [reply .github/workflows/lint.yml:23]: yeah, I did, I just wanted to see if it would help... this should evaporate in a sec
